email-native/src/main/java/org/ballerinalang/stdlib/email/server/EmailListener.java

```diff
@@ -64,25 +67,50 @@ public boolean onMessage(EmailEvent emailEvent) {
         if (runtime != null) {
             Set<Map.Entry<String, ObjectValue>> services = registeredServices.entrySet();
             for (Map.Entry<String, ObjectValue> service : services) {
-                runtime.invokeMethodSync(service.getValue(), ON_MESSAGE, null, ON_MESSAGE_METADATA, email, true);
+                invokeMethodSync(service.getValue(), ON_MESSAGE, ON_MESSAGE_METADATA, email, true);
             }
         } else {
             log.error("Runtime should not be null.");
         }
         return true;
     }
 
+    private void invokeMethodSync(ObjectValue object, String methodName, StrandMetadata metadata, Object... args) {
+        Semaphore semaphore = new Semaphore(0);
+        final ErrorValue[] errorValue = new ErrorValue[1];
+        runtime.invokeMethodAsync(object, methodName, null, metadata, new CallableUnitCallback() {
+            @Override
+            public void notifySuccess() {
+                semaphore.release();
+            }
+
+            @Override
+            public void notifyFailure(ErrorValue error) {
+                errorValue[0] = error;
+                semaphore.release();
+            }
+        }, args);
+        try {
+            semaphore.acquire();
```

Comment:
`invokeMethodAsync` may be required to be called after this line as the Semaphore was not yet acquired at that time.

Reply:
The semaphore is released inside a callback. This is ported from the removed function from the runtime api. If you think async method is more appropriate, its use is recommended.